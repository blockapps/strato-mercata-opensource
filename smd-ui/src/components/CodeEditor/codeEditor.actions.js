export const CODE_EDITOR_COMPILE_REQUEST = "CODE_EDITOR_COMPILE_REQUEST";
export const CODE_EDITOR_COMPILE_SUCCESS = "CODE_EDITOR_COMPILE_SUCCESS";
export const CODE_EDITOR_COMPILE_FAILURE = "CODE_EDITOR_COMPILE_FAILURE";
export const CODE_EDITOR_CHANGE_CREATEACTION = "CODE_EDITOR_CHANGE_CREATEACTION";
export const ADD_NEW_TAB = "ADD_NEW_TAB";
export const REMOVE_TAB = "REMOVE_TAB";
export const ON_TAB_CHANGE = "ON_TAB_CHANGE";
export const EDITOR_CONTRACT_NAME_CHANGE = "EDITOR_CONTRACT_NAME_CHANGE";
export const CHANGE_FILE_NAME = "CHANGE_FILE_NAME";
export const ON_COMPILE_FILE_LOCALLY = "ON_COMPILE_FILE_LOCALLY"

export const changeCreateActionState = function (value, sourceCode, index) {
  return {
    type: CODE_EDITOR_CHANGE_CREATEACTION,
    createActionEnable: value,
    sourceCode: sourceCode,
    index
  }
}

export const compileCodeFromEditor = function (code, codeType) {
  return {
    type: CODE_EDITOR_COMPILE_REQUEST,
    code, 
    codeType
  }
}

export const compileCodeFromEditorSuccess = function (response) {
  return {
    type: CODE_EDITOR_COMPILE_SUCCESS,
    response: response,
  }
}

export const compileCodeFromEditorFailure = function (error) {
  return {
    type: CODE_EDITOR_COMPILE_FAILURE,
    error: error,
  }
}

export const addNewFileTab = function (fileName, text) {
  return {
    type: ADD_NEW_TAB,
    fileName: fileName,
    fileContent: text
  }
}

export const removeTab = function (index) {
  return {
    type: REMOVE_TAB,
    index: index
  }
}

export const onTabChange = function (prevTab, nextTab) {
  return {
    type: ON_TAB_CHANGE,
    prevTab: prevTab,
    nextTab: nextTab
  }
}

export const contractNameChange = function (contractName) {
  return {
    type: EDITOR_CONTRACT_NAME_CHANGE,
    contractName: contractName
  }
}

export const onChangeFileName = function (fileName) {
  return {
    type: CHANGE_FILE_NAME,
    name: fileName
  }
}

export const onCompileFileLocally = function (compileError) {
  return {
    type: ON_COMPILE_FILE_LOCALLY,
    compileError: compileError
  }
}