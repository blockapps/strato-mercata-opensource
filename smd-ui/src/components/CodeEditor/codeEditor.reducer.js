import {
  CODE_EDITOR_COMPILE_REQUEST,
  CODE_EDITOR_COMPILE_SUCCESS,
  CODE_EDITOR_COMPILE_FAILURE,
  CODE_EDITOR_CHANGE_CREATEACTION,
  ADD_NEW_TAB,
  REMOVE_TAB,
  ON_TAB_CHANGE,
  EDITOR_CONTRACT_NAME_CHANGE,
  CHANGE_FILE_NAME,
  ON_COMPILE_FILE_LOCALLY
} from './codeEditor.actions';

const initialState = {
  codeCompileSuccess: undefined,
  abi: undefined,
  fileName: undefined,
  createDisabled: true,
  enableCreateAction: false,
  sourceCode: undefined,
  contractName: undefined,
  tab: [{
    text: '',
    title: 'Main.sol'
  }],
  lastTabSelected: 0,
  currentTabSelected: 0,
  isRemoveTab: false,
  localCompileException: '',
  codeType : "SolidVM"
};

const formatCompilationErrors = function (error) {
  if (typeof error !== "string") {
    return "Failed to connect to the Strato compiler."
  }
  if (error.indexOf('\n') === -1) {
    const jsonErr = JSON.parse(error);
    return jsonErr.replace(/\n/g, "\n");
  }
  let text = error
    .split('\n')
    .reduce((a, part, i) => {
      if (i === 0 || part === '') {
        return a;
      }
      return a + ' ' + part;
    }, '');

  try {
    const jErrors = JSON.parse(text);
    return jErrors.error;
  } catch (e) {
    return text;
  }
}

export const loadState = () => {
  try {
    const serializedState = localStorage.getItem('code_editor_state');
    if (serializedState === null) {
      return initialState;
    }
    return JSON.parse(serializedState);
  } catch (err) {
    return undefined;
  }
};

const reducer = function (state = loadState(), action) {
  switch (action.type) {
    case CODE_EDITOR_COMPILE_REQUEST:
      return {
        ...state,
        response: "Uploading Contract...",
        createDisabled: true,
        enableCreateAction: false,
        codeType : action.codeType
      };

    case CODE_EDITOR_COMPILE_SUCCESS:
      let contracts = action.response && action.response.src && Object.keys(action.response.src);
      return {
        ...state,
        contractName: contracts && contracts[0],
        response: action.response,
        codeCompileSuccess: true,
        enableCreateAction: true
      };

    case CODE_EDITOR_COMPILE_FAILURE:
      return {
        ...state,
        response: formatCompilationErrors(action.error),
        error: action.error,
        codeCompileSuccess: false,
        enableCreateAction: false
      };

    case EDITOR_CONTRACT_NAME_CHANGE:
      return {
        ...state,
        contractName: action.contractName
      }

    case CODE_EDITOR_CHANGE_CREATEACTION:
      const tabItems = [...state.tab];
      tabItems[action.index].text = action.sourceCode;  //new value
      return {
        ...state,
        enableCreateAction: action.createActionEnable,
        sourceCode: action.sourceCode,
        tab: tabItems
      };

    case ADD_NEW_TAB:
      const newTabs = [...state.tab, { title: `${action.fileName}.sol`, text: action.fileContent }]
      return {
        ...state,
        tab: newTabs,
        sourceCode: action.fileContent,
        currentTabSelected: newTabs.length - 1,
        enableCreateAction: false
      };

    case REMOVE_TAB:
      const tabs = state.tab.slice();
      tabs.splice(action.index, 1);
      const selectedTab = action.index === state.currentTabSelected ? (action.index > 0 ? state.currentTabSelected - 1 : 0) : (action.index > state.currentTabSelected ? state.currentTabSelected : state.currentTabSelected - 1)
      return {
        ...state,
        tab: tabs,
        isRemoveTab: true,
        currentTabSelected: selectedTab,
        lastTabSelected: selectedTab
      };

    case ON_TAB_CHANGE:
      const changedTab = state.tab.slice();
      if (state.isRemoveTab) {
        const tabCodeOnRemove = changedTab.length > 0 && changedTab.splice(state.currentTabSelected, 1)
        return {
          ...state,
          isRemoveTab: false,
          sourceCode: tabCodeOnRemove.length > 0 && tabCodeOnRemove[0].text,
          enableCreateAction: false
        }
      }
      const tabCode = changedTab.splice(action.nextTab, 1)
      return {
        ...state,
        sourceCode: tabCode[0].text,
        currentTabSelected: action.nextTab,
        lastTabSelected: action.prevTab,
        enableCreateAction: false
      }

    case CHANGE_FILE_NAME:
      return {
        ...state,
        fileName: action.name
      }

    case ON_COMPILE_FILE_LOCALLY:
      return {
        ...state,
        localCompileException: action.compileError
      }

    default:
      return state;
  }
};

export default reducer;
