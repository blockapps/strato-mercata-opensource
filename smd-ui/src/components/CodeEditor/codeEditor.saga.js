import {
  takeEvery,
  put,
  call
} from 'redux-saga/effects';
import {
  CODE_EDITOR_COMPILE_REQUEST,
  compileCodeFromEditorSuccess,
  compileCodeFromEditorFailure
} from './codeEditor.actions';

import { env } from '../../env';
import { handleErrors } from '../../lib/handleErrors';

const compileUrl = env.BLOC_URL + "/contracts/xabi";
const blocCompileUrl = env.BLOC_URL + "/contracts/compile";

export function tokenizeSource(source) {
  let body = JSON.stringify({src : source});
  return fetch(
    compileUrl,
    {
      method: 'POST',
      credentials: "include",
      headers: {
        "Content-Type": "application/json"
      },
      body 
    })
    .then(function (res) {
      if (res.ok) {
        return res.json();
      } else {
        return res.text().then(function (value) {
          throw value;
        });
      }
    })
    .catch(function (error) {
      throw error;
    });
}

export function compileSource(contractName, source, codeType) {
  const searchable = [];
  return fetch(blocCompileUrl, {
    method: 'POST',
    credentials: "include",
    headers: {
      "accept": "application/json",
      "content-type": "application/json"
    },
    body: JSON.stringify([
      {
        "contractName": contractName,
        "source": source,
        "searchable": searchable,
        "vm" : codeType
      }
    ])
  })
  .then(handleErrors)
  .then(function (res) {
    if (res.ok) {
      return res.json();
    } else {
      return res.text().then(function (value) {
        throw value;
      });
    }
  })
  .then(json => {
    return Promise.resolve(json);
  })
  .catch(function (error) {
    throw error;
  });
}

export function* compileCodeFromEditor(action) {
  try {
    let response
    response = yield call(tokenizeSource, action.code);
    if (response) {
      let contracts = response.src && Object.keys(response.src);
      const contractName = contracts && contracts[0]
      yield call(compileSource, contractName, action.code, action.codeType);
    }
    yield put(compileCodeFromEditorSuccess(response));
  }
  catch (err) {
    yield put(compileCodeFromEditorFailure(err));
  }
}

export function* watchCompileSourceFromEditor() {
  yield takeEvery(CODE_EDITOR_COMPILE_REQUEST, compileCodeFromEditor);
}
