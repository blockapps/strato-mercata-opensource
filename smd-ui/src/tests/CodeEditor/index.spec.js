import React from 'react';
import CodeEditor, { mapStateToProps } from '../../components/CodeEditor/index';
import { extAbi, error, selectedTabContent, codeEditor, sourceCodeUndefinedImport } from './codeEditorMock';
import { createStore, combineReducers } from 'redux'
import { reducer as formReducer } from 'redux-form'

describe('CodeEditor: index', () => {
  let files
  const readAsText = jest.fn();
  let store

  beforeEach(() => {
    store = createStore(combineReducers({ form: formReducer }))
  })

  beforeAll(() => {
    localStorage.clear();
    files = [
      {
        name: 'file1.sol',
        size: 1111,
        type: ''
      }
    ]
    const fileContents = 'file contents';
    const expectedFinalState = { fileContents: fileContents };
    const file = new Blob([fileContents], { type: 'text/plain' });
    const addEventListener = jest.fn((_, evtHandler) => { evtHandler(); });
    const dummyFileReader = { addEventListener, readAsText, result: fileContents };
    window.FileReader = jest.fn(() => dummyFileReader);
  });

  test('render codeEditor with default values', () => {
    const props = {
      selectedTabContent: selectedTabContent,
      codeEditorData: codeEditor
    }
    const wrapper = shallow(
      <CodeEditor.WrappedComponent {...props} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('mapStateToProps with default values', () => {
    const state = {
      codeEditor: codeEditor,
      chains: {
        selectedChain: "abcdefg",
        chainIds: [],
      }
    }
    expect(mapStateToProps(state)).toMatchSnapshot();
  });

  describe('simulate events', () => {

    test('compile SolidVM code', () => {
      const props = {
        selectedTabContent: selectedTabContent,
        codeEditorData: codeEditor,
        onCompileFileLocally: jest.fn(),
        removeTab: jest.fn(),
        changeCreateActionState: jest.fn(),
        onTabChange: jest.fn(),
        onChangeFileName: jest.fn(),
        addNewFileTab: jest.fn(),
        fetchChainDetailSelect: jest.fn(),
        selectChain: jest.fn(),
        fetchChainIds: jest.fn(),
        selectedChain: "abcdefg",
        chainIds: ["abcdefg"],
        store,
      }
      const wrapper = shallow(
        <CodeEditor.WrappedComponent {...props} />
      ).dive().dive().dive();
      const popOver = wrapper.find('Popover').at(0)
      wrapper.find('Button').at(0).simulate('click')
      expect(props.onCompileFileLocally).toHaveBeenCalled()
      wrapper.find('Button').at(1).simulate('click')
      wrapper.find('Button').at(2).simulate('click')
      wrapper.find('Button').at(3).simulate('click')
  
      wrapper.find('span').at(0).simulate('click')
      expect(props.removeTab).toHaveBeenCalled()
  
      const monacoEditor = wrapper.find('Tab2').at(1).dive().find('MonacoEditor')
      monacoEditor.simulate('change', { target: { value: 'abc', index: 1 } })
      expect(props.changeCreateActionState).toHaveBeenCalled()
  
      wrapper.find('Tabs2').at(0).simulate('change', { target: { newTab: 1, prevTab: 0 } })
      expect(props.onTabChange).toHaveBeenCalled()
  
      const input = popOver.dive().find('input')
      expect(input.value).toBe(undefined)
      input.simulate('change', { target: { value: 'tanuj' } })
      expect(props.onChangeFileName).toHaveBeenCalled()
  
      const button = popOver.dive().find('Overlay').find('Button')
      button.simulate('click')
      expect(props.addNewFileTab).toHaveBeenCalled()
  
      props.codeEditorData['sourceCode'] = undefined
      wrapper.find('Button').at(0).simulate('click')
      expect(props.onCompileFileLocally).toHaveBeenCalled()
  
      props.codeEditorData['sourceCode'] = sourceCodeUndefinedImport
      wrapper.find('Button').at(0).simulate('click')
      expect(props.onCompileFileLocally).toHaveBeenCalled()
    })

    // const props = {
    //   selectedTabContent: selectedTabContent,
    //   codeEditorData: codeEditor,
    //   onCompileFileLocally: jest.fn(),
    //   removeTab: jest.fn(),
    //   changeCreateActionState: jest.fn(),
    //   onTabChange: jest.fn(),
    //   onChangeFileName: jest.fn(),
    //   addNewFileTab: jest.fn(),
    //   // selectedChain: "abcdefg",
    //   // chainIds: ["abcdefg"],
    //   // chains: {}
    // }
    // const wrapper = shallow(
    //   <CodeEditor.WrappedComponent {...props} />
    // );
    

    // wrapper.find('span').at(0).simulate('click')
    // expect(props.removeTab).toHaveBeenCalled()

    // const monacoEditor = wrapper.find('Tab2').at(1).dive().find('MonacoEditor')
    // monacoEditor.simulate('change', { target: { value: 'abc', index: 1 } })
    // expect(props.changeCreateActionState).toHaveBeenCalled()

    // wrapper.find('Tabs2').at(0).simulate('change', { target: { newTab: 1, prevTab: 0 } })
    // expect(props.onTabChange).toHaveBeenCalled()

    // const input = popOver.dive().find('input')
    // expect(input.value).toBe(undefined)
    // input.simulate('change', { target: { value: 'tanuj' } })
    // expect(props.onChangeFileName).toHaveBeenCalled()

    // const button = popOver.dive().find('Overlay').find('Button')
    // button.simulate('click')
    // expect(props.addNewFileTab).toHaveBeenCalled()

    // props.codeEditorData['sourceCode'] = undefined
    // wrapper.find('Button').at(0).simulate('click')
    // expect(props.onCompileFileLocally).toHaveBeenCalled()

    // props.codeEditorData['sourceCode'] = sourceCodeUndefinedImport
    // wrapper.find('Button').at(0).simulate('click')
    // expect(props.onCompileFileLocally).toHaveBeenCalled()

  });

  test('.sol file drop', () => {
    const props = {
      selectedTabContent: selectedTabContent,
      codeEditorData: codeEditor,
      onCompileFileLocally: jest.fn(),
      removeTab: jest.fn(),
      changeCreateActionState: jest.fn(),
      onTabChange: jest.fn(),
      onChangeFileName: jest.fn(),
      addNewFileTab: jest.fn(),
      fetchChainDetailSelect: jest.fn(),
      selectChain: jest.fn(),
      fetchChainIds: jest.fn(),
      selectedChain: "adfads",
      chainIds: ["adfads"],
      store,
    }
    const wrapper = shallow(
      <CodeEditor.WrappedComponent {...props} />
    ).dive().dive().dive();
    console.log(wrapper)
    const dropzone = wrapper.find('.dropzone')
    console.log(dropzone)
    dropzone.simulate('drop', files)
    expect(readAsText).toHaveBeenCalled()

  });

  test('other file eg. text file drop', () => {
    const props = {
      selectedTabContent: selectedTabContent,
      codeEditorData: codeEditor,
      onCompileFileLocally: jest.fn(),
      removeTab: jest.fn(),
      changeCreateActionState: jest.fn(),
      onTabChange: jest.fn(),
      onChangeFileName: jest.fn(),
      addNewFileTab: jest.fn(),
      fetchChainDetailSelect: jest.fn(),
      selectChain: jest.fn(),
      fetchChainIds: jest.fn(),
      selectedChain: "adfads",
      chainIds: ["adfads"],
      store,
    }
    const wrapper = shallow(
      <CodeEditor.WrappedComponent {...props} />
    ).dive().dive().dive();
    files[0]['name'] = 'pqr.txt'
    const dropzone = wrapper.find('Dropzone')
    dropzone.simulate('drop', files)
    expect(readAsText).toHaveBeenCalled()
  });

  test('multiple file drop', () => {
    const props = {
      selectedTabContent: selectedTabContent,
      codeEditorData: codeEditor,
      onCompileFileLocally: jest.fn(),
      removeTab: jest.fn(),
      changeCreateActionState: jest.fn(),
      onTabChange: jest.fn(),
      onChangeFileName: jest.fn(),
      addNewFileTab: jest.fn(),
      fetchChainDetailSelect: jest.fn(),
      selectChain: jest.fn(),
      fetchChainIds: jest.fn(),
      selectedChain: "adfads",
      chainIds: ["adfads"],
      store,
    }
    const wrapper = shallow(
      <CodeEditor.WrappedComponent {...props} />
    ).dive().dive().dive();
    files.push({
      name: 'file2.sol',
      size: 1111,
      type: ''
    })
    const dropzone = wrapper.find('Dropzone')
    dropzone.simulate('drop', files)
    expect(readAsText).toHaveBeenCalled()
  });

  test('local storage', () => {
    const props = {
      selectedTabContent: selectedTabContent,
      codeEditorData: codeEditor,
      onCompileFileLocally: jest.fn(),
      removeTab: jest.fn(),
      changeCreateActionState: jest.fn(),
      onTabChange: jest.fn(),
      onChangeFileName: jest.fn(),
      addNewFileTab: jest.fn(),
      fetchChainDetailSelect: jest.fn(),
      selectChain: jest.fn(),
      fetchChainIds: jest.fn(),
      selectedChain: "adfads",
      chainIds: ["adfads"],
      store,
    }
    var wrapper = shallow(
      <CodeEditor.WrappedComponent {...props} />, { lifecycleExperimental: true }
    ).dive().dive().dive();
    window.onbeforeunload()
    expect(Object.keys(localStorage.__STORE__).length).toBe(1)
  });

  test('save local state method call', () => {
    const props = {
      selectedTabContent: selectedTabContent,
      codeEditorData: codeEditor,
      onCompileFileLocally: jest.fn(),
      removeTab: jest.fn(),
      changeCreateActionState: jest.fn(),
      onTabChange: jest.fn(),
      onChangeFileName: jest.fn(),
      addNewFileTab: jest.fn(),
      fetchChainDetailSelect: jest.fn(),
      selectChain: jest.fn(),
      fetchChainIds: jest.fn(),
      selectedChain: "adfads",
      chainIds: ["adfads"],
      store,
    }
    const options = {
      lifecycleExperimental: true,
      disableLifecycleMethods: false
    };
    var wrapper = shallow(
      <CodeEditor.WrappedComponent {...props} />, options
    ).dive().dive().dive();
    wrapper.instance().saveLocalState = jest.fn()
    wrapper.instance().componentDidUpdate()
    expect(wrapper.instance().saveLocalState).toHaveBeenCalled()
  });

})