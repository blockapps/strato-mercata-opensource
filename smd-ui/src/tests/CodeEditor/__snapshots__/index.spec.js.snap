// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeEditor: index mapStateToProps with default values 1`] = `
Object {
  "chainIds": Array [],
  "codeEditorData": Object {
    "codeCompileSuccess": false,
    "contractName": "Cloner",
    "createDisabled": true,
    "currentTabSelected": 0,
    "enableCreateAction": false,
    "error": "src:6:77: Error: Expected token Semicolon got 'RBrace'
{ Sheep sheep = new Sheep(name, sheepDNA); clones.push(sheep); return sheep }
                                                                            ^
",
    "fileName": "SampleStorage",
    "isRemoveTab": false,
    "lastTabSelected": 0,
    "localCompileException": "",
    "response": "src:6:77: Error: Expected token Semicolon got 'RBrace'↵{ Sheep sheep = new Sheep(name, sheepDNA); clones.push(sheep); return sheep }↵                                                                            ^↵",
    "serverText": "HelloFri Nov 17 2017 15:51:33 GMT+0530 (IST)",
    "sourceCode": "contract Cloner {↵Sheep[] public clones;↵string public sheepDNA = \\"ccc\\";↵↵function cloneSheep(string name) returns (address)↵{ Sheep sheep = new Sheep(name, sheepDNA); clones.push(sheep); return sheep }↵↵}↵↵contract Sheep {↵string public name;↵string public dna;↵↵function Sheep(string _name, string _dna)↵{ name = _name; dna = _dna; }↵↵function geneticallyModify(string _dna)↵{ dna = _dna; }↵↵}",
    "tab": Array [
      Object {
        "text": "contract Cloner {↵Sheep[] public clones;↵string public sheepDNA = \\"ccc\\";↵↵function cloneSheep(string name) returns (address)↵{ Sheep sheep = new Sheep(name, sheepDNA); clones.push(sheep); return sheep }↵↵}↵↵contract Sheep {↵string public name;↵string public dna;↵↵function Sheep(string _name, string _dna)↵{ name = _name; dna = _dna; }↵↵function geneticallyModify(string _dna)↵{ dna = _dna; }↵↵}",
        "title": "abc.sol",
      },
      Object {
        "text": "contract Cloner {↵Sheep[] public clones;↵string public sheepDNA = \\"ccc\\";↵↵function cloneSheep(string name) returns (address)↵{ Sheep sheep = new Sheep(name, sheepDNA); clones.push(sheep); return sheep }↵↵}↵↵contract Sheep {↵string public name;↵string public dna;↵↵function Sheep(string _name, string _dna)↵{ name = _name; dna = _dna; }↵↵function geneticallyModify(string _dna)↵{ dna = _dna; }↵↵}",
        "title": "xyz.sol",
      },
    ],
    "timer": 87,
  },
  "selectedChain": "abcdefg",
  "selectedTabContent": Object {
    "text": "contract Cloner {↵Sheep[] public clones;↵string public sheepDNA = \\"ccc\\";↵↵function cloneSheep(string name) returns (address)↵{ Sheep sheep = new Sheep(name, sheepDNA); clones.push(sheep); return sheep }↵↵}↵↵contract Sheep {↵string public name;↵string public dna;↵↵function Sheep(string _name, string _dna)↵{ name = _name; dna = _dna; }↵↵function geneticallyModify(string _dna)↵{ dna = _dna; }↵↵}",
    "title": "abc.sol",
  },
}
`;

exports[`CodeEditor: index render codeEditor with default values 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ReduxForm
    codeEditorData={
      Object {
        "codeCompileSuccess": false,
        "contractName": "Cloner",
        "createDisabled": true,
        "currentTabSelected": 0,
        "enableCreateAction": false,
        "error": "src:6:77: Error: Expected token Semicolon got 'RBrace'
{ Sheep sheep = new Sheep(name, sheepDNA); clones.push(sheep); return sheep }
                                                                            ^
",
        "fileName": "SampleStorage",
        "isRemoveTab": false,
        "lastTabSelected": 0,
        "localCompileException": "",
        "response": "src:6:77: Error: Expected token Semicolon got 'RBrace'↵{ Sheep sheep = new Sheep(name, sheepDNA); clones.push(sheep); return sheep }↵                                                                            ^↵",
        "serverText": "HelloFri Nov 17 2017 15:51:33 GMT+0530 (IST)",
        "sourceCode": "contract Cloner {↵Sheep[] public clones;↵string public sheepDNA = \\"ccc\\";↵↵function cloneSheep(string name) returns (address)↵{ Sheep sheep = new Sheep(name, sheepDNA); clones.push(sheep); return sheep }↵↵}↵↵contract Sheep {↵string public name;↵string public dna;↵↵function Sheep(string _name, string _dna)↵{ name = _name; dna = _dna; }↵↵function geneticallyModify(string _dna)↵{ dna = _dna; }↵↵}",
        "tab": Array [
          Object {
            "text": "contract Cloner {↵Sheep[] public clones;↵string public sheepDNA = \\"ccc\\";↵↵function cloneSheep(string name) returns (address)↵{ Sheep sheep = new Sheep(name, sheepDNA); clones.push(sheep); return sheep }↵↵}↵↵contract Sheep {↵string public name;↵string public dna;↵↵function Sheep(string _name, string _dna)↵{ name = _name; dna = _dna; }↵↵function geneticallyModify(string _dna)↵{ dna = _dna; }↵↵}",
            "title": "abc.sol",
          },
          Object {
            "text": "contract Cloner {↵Sheep[] public clones;↵string public sheepDNA = \\"ccc\\";↵↵function cloneSheep(string name) returns (address)↵{ Sheep sheep = new Sheep(name, sheepDNA); clones.push(sheep); return sheep }↵↵}↵↵contract Sheep {↵string public name;↵string public dna;↵↵function Sheep(string _name, string _dna)↵{ name = _name; dna = _dna; }↵↵function geneticallyModify(string _dna)↵{ dna = _dna; }↵↵}",
            "title": "xyz.sol",
          },
        ],
        "timer": 87,
      }
    }
    selectedTabContent={
      Object {
        "text": "contract Cloner {↵Sheep[] public clones;↵string public sheepDNA = \\"ccc\\";↵↵function cloneSheep(string name) returns (address)↵{ Sheep sheep = new Sheep(name, sheepDNA); clones.push(sheep); return sheep }↵↵}↵↵contract Sheep {↵string public name;↵string public dna;↵↵function Sheep(string _name, string _dna)↵{ name = _name; dna = _dna; }↵↵function geneticallyModify(string _dna)↵{ dna = _dna; }↵↵}",
        "title": "abc.sol",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "codeEditorData": Object {
        "codeCompileSuccess": false,
        "contractName": "Cloner",
        "createDisabled": true,
        "currentTabSelected": 0,
        "enableCreateAction": false,
        "error": "src:6:77: Error: Expected token Semicolon got 'RBrace'
{ Sheep sheep = new Sheep(name, sheepDNA); clones.push(sheep); return sheep }
                                                                            ^
",
        "fileName": "SampleStorage",
        "isRemoveTab": false,
        "lastTabSelected": 0,
        "localCompileException": "",
        "response": "src:6:77: Error: Expected token Semicolon got 'RBrace'↵{ Sheep sheep = new Sheep(name, sheepDNA); clones.push(sheep); return sheep }↵                                                                            ^↵",
        "serverText": "HelloFri Nov 17 2017 15:51:33 GMT+0530 (IST)",
        "sourceCode": "contract Cloner {↵Sheep[] public clones;↵string public sheepDNA = \\"ccc\\";↵↵function cloneSheep(string name) returns (address)↵{ Sheep sheep = new Sheep(name, sheepDNA); clones.push(sheep); return sheep }↵↵}↵↵contract Sheep {↵string public name;↵string public dna;↵↵function Sheep(string _name, string _dna)↵{ name = _name; dna = _dna; }↵↵function geneticallyModify(string _dna)↵{ dna = _dna; }↵↵}",
        "tab": Array [
          Object {
            "text": "contract Cloner {↵Sheep[] public clones;↵string public sheepDNA = \\"ccc\\";↵↵function cloneSheep(string name) returns (address)↵{ Sheep sheep = new Sheep(name, sheepDNA); clones.push(sheep); return sheep }↵↵}↵↵contract Sheep {↵string public name;↵string public dna;↵↵function Sheep(string _name, string _dna)↵{ name = _name; dna = _dna; }↵↵function geneticallyModify(string _dna)↵{ dna = _dna; }↵↵}",
            "title": "abc.sol",
          },
          Object {
            "text": "contract Cloner {↵Sheep[] public clones;↵string public sheepDNA = \\"ccc\\";↵↵function cloneSheep(string name) returns (address)↵{ Sheep sheep = new Sheep(name, sheepDNA); clones.push(sheep); return sheep }↵↵}↵↵contract Sheep {↵string public name;↵string public dna;↵↵function Sheep(string _name, string _dna)↵{ name = _name; dna = _dna; }↵↵function geneticallyModify(string _dna)↵{ dna = _dna; }↵↵}",
            "title": "xyz.sol",
          },
        ],
        "timer": 87,
      },
      "destroyOnUnmount": true,
      "enableReinitialize": false,
      "forceUnregisterOnUnmount": false,
      "form": "CodeEditor",
      "getFormState": [Function],
      "initialValues": undefined,
      "keepDirtyOnReinitialize": false,
      "persistentSubmitErrors": false,
      "pure": true,
      "selectedTabContent": Object {
        "text": "contract Cloner {↵Sheep[] public clones;↵string public sheepDNA = \\"ccc\\";↵↵function cloneSheep(string name) returns (address)↵{ Sheep sheep = new Sheep(name, sheepDNA); clones.push(sheep); return sheep }↵↵}↵↵contract Sheep {↵string public name;↵string public dna;↵↵function Sheep(string _name, string _dna)↵{ name = _name; dna = _dna; }↵↵function geneticallyModify(string _dna)↵{ dna = _dna; }↵↵}",
        "title": "abc.sol",
      },
      "shouldAsyncValidate": [Function],
      "shouldValidate": [Function],
      "touchOnBlur": true,
      "touchOnChange": false,
    },
    "ref": "wrapped",
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "codeEditorData": Object {
          "codeCompileSuccess": false,
          "contractName": "Cloner",
          "createDisabled": true,
          "currentTabSelected": 0,
          "enableCreateAction": false,
          "error": "src:6:77: Error: Expected token Semicolon got 'RBrace'
{ Sheep sheep = new Sheep(name, sheepDNA); clones.push(sheep); return sheep }
                                                                            ^
",
          "fileName": "SampleStorage",
          "isRemoveTab": false,
          "lastTabSelected": 0,
          "localCompileException": "",
          "response": "src:6:77: Error: Expected token Semicolon got 'RBrace'↵{ Sheep sheep = new Sheep(name, sheepDNA); clones.push(sheep); return sheep }↵                                                                            ^↵",
          "serverText": "HelloFri Nov 17 2017 15:51:33 GMT+0530 (IST)",
          "sourceCode": "contract Cloner {↵Sheep[] public clones;↵string public sheepDNA = \\"ccc\\";↵↵function cloneSheep(string name) returns (address)↵{ Sheep sheep = new Sheep(name, sheepDNA); clones.push(sheep); return sheep }↵↵}↵↵contract Sheep {↵string public name;↵string public dna;↵↵function Sheep(string _name, string _dna)↵{ name = _name; dna = _dna; }↵↵function geneticallyModify(string _dna)↵{ dna = _dna; }↵↵}",
          "tab": Array [
            Object {
              "text": "contract Cloner {↵Sheep[] public clones;↵string public sheepDNA = \\"ccc\\";↵↵function cloneSheep(string name) returns (address)↵{ Sheep sheep = new Sheep(name, sheepDNA); clones.push(sheep); return sheep }↵↵}↵↵contract Sheep {↵string public name;↵string public dna;↵↵function Sheep(string _name, string _dna)↵{ name = _name; dna = _dna; }↵↵function geneticallyModify(string _dna)↵{ dna = _dna; }↵↵}",
              "title": "abc.sol",
            },
            Object {
              "text": "contract Cloner {↵Sheep[] public clones;↵string public sheepDNA = \\"ccc\\";↵↵function cloneSheep(string name) returns (address)↵{ Sheep sheep = new Sheep(name, sheepDNA); clones.push(sheep); return sheep }↵↵}↵↵contract Sheep {↵string public name;↵string public dna;↵↵function Sheep(string _name, string _dna)↵{ name = _name; dna = _dna; }↵↵function geneticallyModify(string _dna)↵{ dna = _dna; }↵↵}",
              "title": "xyz.sol",
            },
          ],
          "timer": 87,
        },
        "destroyOnUnmount": true,
        "enableReinitialize": false,
        "forceUnregisterOnUnmount": false,
        "form": "CodeEditor",
        "getFormState": [Function],
        "initialValues": undefined,
        "keepDirtyOnReinitialize": false,
        "persistentSubmitErrors": false,
        "pure": true,
        "selectedTabContent": Object {
          "text": "contract Cloner {↵Sheep[] public clones;↵string public sheepDNA = \\"ccc\\";↵↵function cloneSheep(string name) returns (address)↵{ Sheep sheep = new Sheep(name, sheepDNA); clones.push(sheep); return sheep }↵↵}↵↵contract Sheep {↵string public name;↵string public dna;↵↵function Sheep(string _name, string _dna)↵{ name = _name; dna = _dna; }↵↵function geneticallyModify(string _dna)↵{ dna = _dna; }↵↵}",
          "title": "abc.sol",
        },
        "shouldAsyncValidate": [Function],
        "shouldValidate": [Function],
        "touchOnBlur": true,
        "touchOnChange": false,
      },
      "ref": "wrapped",
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
